source/agora/test/EnrollmentManager.d

```diff
@@ -142,3 +142,82 @@ unittest
     assert(b20.header.enrollments.length == 1);
     assert(b20.header.enrollments[0] == enroll);
 }
+
+/// Situation: A node in the network crashes after a new enrollment for the
+///     node has been inserted into a new block with consensus. And the node
+///     starts again
+/// Expectation: The node restores its own enrollment from the chain and
+///     reveals its pre-images preidically.
+unittest
+{
+    // Boilerplate
+    TestConf conf = {
+        validator_cycle : 20,
+    };
+    auto network = makeTestNetwork(conf);
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.start();
+    network.waitForDiscovery();
+    auto nodes = network.clients;
+
+    // Sanity check: Check if genesis block has enrollments
+    const b0 = nodes[0].getBlocksFrom(0, 2)[0];
+    assert(b0.header.enrollments.length >= 1);
+
+    // Make 19 blocks
+    Transaction[] txs;
+    foreach (count; 0 .. 19)
+    {
+        // Use Genesis
+        if (!txs.length)
+            txs = genesisSpendable().take(Block.TxsInBlock).enumerate()
+                .map!(en => en.value.refund(WK.Keys.A.address).sign())
+                .array();
+        else
+            txs = txs
+                .map!(ptx => TxBuilder(ptx).refund(WK.Keys[count].address).sign())
+                .array();
+        txs.each!(tx => nodes[1].putTransaction(tx));
+
+        // Ensure everyone is at the same level
+        ensureConsistency(nodes, count + 1);
+    }
+
+    // Now create an Enrollment for nodes[0], create block #20, and restart nodes[0]
+    auto enroll = nodes[0].createEnrollmentData();
+    nodes[0].enrollValidator(enroll);
+    txs = txs
+        .map!(ptx => TxBuilder(ptx).refund(WK.Keys[20].address).sign())
+        .array();
+    txs.each!(tx => nodes[1].putTransaction(tx));
+    ensureConsistency(nodes.take(1), 20);
+    assert(nodes[0].getBlocksFrom(20, 1)[0].header.enrollments.length == 1);
```

Comment:
I think this is better with `retryFor` in case the request times out (it means there will be less random test-suite failures).